src/backend/gl/src/lib.rs

```diff
@@ -484,35 +490,34 @@ impl hal::PhysicalDevice<Backend> for PhysicalDevice {
         // COHERENT flags require that the backend does flushing and invalidation
         // by itself. If we move towards persistent mapping we need to re-evaluate it.
         let caps = &self.0.private_caps;
-        let memory_types = if caps.map || caps.emulate_map {
-            vec![
-                hal::MemoryType {
-                    properties: Properties::DEVICE_LOCAL,
-                    heap_index: 1,
-                },
-                hal::MemoryType {
-                    // upload
-                    properties: Properties::CPU_VISIBLE | Properties::COHERENT,
-                    heap_index: 0,
-                },
-                hal::MemoryType {
-                    // download
-                    properties: Properties::CPU_VISIBLE
-                        | Properties::COHERENT
-                        | Properties::CPU_CACHED,
-                    heap_index: 0,
-                },
-            ]
-        } else {
-            vec![hal::MemoryType {
+        assert!(caps.map || caps.emulate_map);
+        let memory_types = vec![
+            // Faked DEVICE_LOCAL memory for images, no gl buffer is actually allocated for it.
+            hal::MemoryType {
                 properties: Properties::DEVICE_LOCAL,
                 heap_index: 0,
-            }]
-        };
+            },
+            // Memory type for uses other than images and INDEX
+            hal::MemoryType {
```

Comment:
would be good to also expose a device-local memory type for buffers

Reply:
That sounded right to me too, but I didn't know enough to know how to do it correctly.  If we have a device-local buffer type, what should the gl backend do when copying from a cpu-visible buffer (backed by a real native buffer) to a device-local buffer?  I imagine the correct answer is *nothing*, but what happens then when the user tries to copy from a fake device-local buffer to e.g. an image buffer?  This might be another case where if I knew more about how HAL works it would make more sense to me, but as of right now I need it explained a bit slower :D

Reply:
Well, now that I think about it I suppose just having an actual, real buffer without any memory mapping would be the right thing to do, even if it needlessly duplicates memory on platforms with emulated memory mapping?

Reply:
Okay I think I pushed something that works, it at least fixes the wgpu-rs shadow example

Reply:
Copying from one buffer to another would be going through `glCopyBufferSubData`.

>what happens then when the user tries to copy from a fake device-local buffer to e.g. an image buffer

Device-local buffer is just a GL buffer (or a sub-range of it, if we have a buffer associated with `Memory`) without MAP_XXX flags. Copying from it, to it, using as index/vertex, all those things are done the same way as usual.

Reply:
Yeah, that's where I ended up.